### PR TITLE
Typings updates

### PR DESCRIPTION
**Pull request by koskimas**
I added new methods and fixed some errors along the way.

Biggest change is using a new `AnyObject` interface in model hooks. Previously they used `object` which made overriding them really difficult since you cannot access any property through `object`. If there is a better way to achieve the same, feel free to point it out.

### CODE REVIEW

File: typings/objection/index.d.ts [hunk 1]
```diff
@@ -14,10 +14,17 @@ declare namespace Objection {
...
 
+  const snakeCaseMappers: () => ColumnNameMappers;
+  const knexSnakeCaseMappers: () => KnexMappers;
+
   interface LiteralObject {
     [key: string]: Value;
   }
 
+  export interface AnyObject {
+    [key: string]: any;
```
Comment (mceachen) [inline typings/objection/index.d.ts]: Should the value here really be any? Like, is it OK to have an array of model objects (not just object representations of model objects, but actual instances of Model subclasses)? It should nest other instances, correct?

See https://github.com/Vincit/objection.js/issues/558#issue-265887726 for a more constrained "AnyObject" definition, if we need that.

I think we can do better with the name, too, given that "object" already means two things in [TypeScript](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type) (with `object` and `Object`) and `any` is a superset of those, so it's compounding two names with three definitions. 

Is `SimpleObject` better?

```ts
  type SimpleValue = string | number | boolean | Date | null | undefined | SimpleObject 

  interface SimpleObject {
    [key: string]: SimpleValue | SimpleValue[]
  }
```

File: typings/objection/index.d.ts [hunk 2]
```diff
@@ -622,9 +651,10 @@ declare namespace Objection {
     isExecutable(): boolean;
     isFindQuery(): boolean;
 
-    runBefore(fn: (builder: this) => void): this;
+    runBefore(fn: (result: any, builder: this) => any): this;
```
Comment (mceachen) [inline typings/objection/index.d.ts]: Is `any` the best we can do here for these?

Comment (koskimas) [reply typings/objection/index.d.ts]: For runBefore the value can literally be anything. runAfter gets passed the query result, so we could improve that, but we would need to move the method to each of the different QueryBuilder classes. Return value can be anything.

File: typings/objection/index.d.ts (hunk 1 above)
Comment (koskimas) [reply typings/objection/index.d.ts]: Usually the objects are JSON objects, but they can be basically anything. Database drivers can have their own conversions and users can pass anything to `fromJson` etc. Also I don't really see the use of limiting that type. Can you give me an example where that would be useful?

File: tests/ts/examples.ts [hunk 3]
```diff
@@ -24,6 +27,16 @@ class Person extends objection.Model {
   async $beforeInsert(queryContext: objection.QueryContext) {
     console.log(queryContext.someCustomValue);
   }
+
+  $formatDatabaseJson(json: objection.AnyObject) {
+    json.bar = json.foo;
```
[the remark this answers is not in the document]
Comment (koskimas) [reply tests/ts/examples.ts]: I'll do that

File: tests/ts/examples.ts [hunk 4]
```diff
@@ -13,6 +13,9 @@ const {lit, raw, ref} = objection;
 class Person extends objection.Model {
   firstName: string;
   lastName: string;
+
+  static columnNameMappers = objection.snakeCaseMappers();
```
[the remark this answers is not in the document]
Comment (koskimas) [reply tests/ts/examples.ts]: I'll add a "test" that uses the return value.

File: typings/objection/index.d.ts (hunk 1 above)
Comment (koskimas) [reply typings/objection/index.d.ts]: You are right about the name, but I don't think SimpleObject is better. The object may be very complex.

Comment (koskimas) [reply typings/objection/index.d.ts]: What about `Pojo`?

Comment (koskimas) [reply typings/objection/index.d.ts]: or `PlainObject`?

Comment (mceachen) [reply typings/objection/index.d.ts]: "Pojo" is perfect.

File: typings/objection/index.d.ts (hunk 2 above)
Comment (mceachen) [reply typings/objection/index.d.ts]: ok!

Comment (koskimas) [reply typings/objection/index.d.ts]: I fixed the `runAfter` `result` type.

**Comment by koskimas** [comment]
@mceachen Is there still something you'd like to change or can I merge this?

**Comment by mceachen** [comment]
Looks good to me!

**Pull request merged**